packages/plugin-drizzle/src/index.ts

```diff
@@ -1,5 +1,242 @@
+      const cursor = request.startingCursor;
+
+      if (!cursor) {
+        return;
+      }
+
+      await withTransaction(db, async (tx) => {
+        await invalidate(tx, cursor, idColumn);
+
+        if (enablePersistence) {
+          await invalidateState({ tx, cursor, indexerName });
+        }
+      });
+    });
+
+    indexer.hooks.hook("connect:factory", async ({ request, endCursor }) => {
+      if (!enablePersistence) {
+        return;
+      }
+      // We can call this hook because this hook is called inside the transaction of handler:middleware
+      // so we have access to the transaction from the context
+      const { db: tx } = useDrizzleStorage(db);
+
+      if (endCursor && request.filter[1]) {
+        await persistState({
+          tx,
+          endCursor,
+          filter: request.filter[1],
+          indexerName,
+        });
+      }
+    });
+
+    indexer.hooks.hook("message:finalize", async ({ message }) => {
+      const { cursor } = message.finalize;
+
+      if (!cursor) {
+        throw new DrizzleStorageError("finalized cursor is undefined");
+      }
+
+      await withTransaction(db, async (tx) => {
+        await finalize(tx, cursor);
+
+        if (enablePersistence) {
+          await finalizeState({ tx, cursor, indexerName });
+        }
+      });
+    });
+
+    indexer.hooks.hook("message:invalidate", async ({ message }) => {
+      const { cursor } = message.invalidate;
+
+      if (!cursor) {
+        throw new DrizzleStorageError("invalidate cursor is undefined");
+      }
+
+      await withTransaction(db, async (tx) => {
+        await invalidate(tx, cursor, idColumn);
+
+        if (enablePersistence) {
+          await invalidateState({ tx, cursor, indexerName });
+        }
+      });
+    });
+
+    indexer.hooks.hook("handler:middleware", async ({ use }) => {
+      use(async (context, next) => {
+        const { endCursor, finality } = context as {
+          endCursor: Cursor;
+          finality: DataFinality;
+        };
+
+        if (!endCursor) {
+          throw new DrizzleStorageError("end cursor is undefined");
+        }
+
+        await withTransaction(db, async (tx) => {
+          context[DRIZZLE_PROPERTY] = { db: tx } as DrizzleStorage<
+            TQueryResult,
+            TFullSchema,
+            TSchema
+          >;
+
+          if (finality !== "finalized") {
+            await registerTriggers(tx, tableNames, endCursor, idColumn);
+          }
+
+          await next();
+          delete context[DRIZZLE_PROPERTY];
+
+          if (enablePersistence) {
+            await persistState({
+              tx,
+              endCursor,
+              indexerName,
+            });
+          }
+        });
+
+        if (finality !== "finalized") {
+          // remove trigger outside of the transaction or it won't be triggered.
+          await removeTriggers(db, tableNames);
+        }
```

Comment:
Probably want to wrap this in a `try { ... } catch (err) { throw err }` to remove the triggers in all cases.